### ISSUE TITLE
Including non-constant time annotations for dynamic execution analysis

### ISSUE THREAD

**Issue by Voker57** [openssl/openssl#8912]
Hello

Our research team have made a execution analysis tool called Triggerflow and applied it to OpenSSL, tracking down several bugs in usage of BN_FLG_CONSTTIME leading to leaking timing information. We are currently maintain our annotations (comments denoting code that can leak timing info about private data) as patches, incorporating them into OpenSSL upstream would greatly reduce complexity of testing.

Our current annotation set: https://gitlab.com/nisec/openssl-triggerflow-ci/tree/master/patches
Paper: https://eprint.iacr.org/2019/366.pdf
Tool: https://gitlab.com/nisec/triggerflow
Our CI setup monitoring live master, 1.1.1 and 1.1.0 branches: https://gitlab.com/nisec/openssl-triggerflow-ci

**Comment by paulidale** [comment]
Thoughts @openssl/committers on including these tags?

**Comment by bbbrumley** [comment]
If merged here upstream, we can also extend the annotations to be more informative, so they are not just tags. e.g. instead of

```
 int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
 {
-    BIGNUM *a, *b, *t;
+    BIGNUM *a, *b, *t; /* TRIGGERFLOW_POI */
```

Some documentation

```
 int BN_gcd(BIGNUM *r, const BIGNUM *in_a, const BIGNUM *in_b, BN_CTX *ctx)
 {
    BIGNUM *a, *b, *t;
    /*-
     * This function calls "euclid", that branches on the input.
     * If "in_a" or "in_b" are secret, they are potentially exposed
     * to side-channel capable attackers.
     *
     * TRIGGERFLOW_POI
     */
```

To provide some additional value outside of automated testing.

**Comment by kroeckx** [comment]
If we start to add such comments, I think this also opens the door for other annotations to be added that we currently don't do for static or dynamic checking of the code. This might results in a lot of comments being added for various tools.

I think we should at least look at this case per case, to see what the costs and benefits are.

I think we should add triggerflow, and run it in CI. And comments like above are very useful.

**Comment by mattcaswell** [comment]
How extensive are these comments likely to be?

**Comment by bbbrumley** [comment]
> How extensive are these comments likely to be?

Here's our [current patchset](https://gitlab.com/nisec/openssl-triggerflow-ci/tree/master/patches).

**Comment by paulidale** [comment]
I'd approve the extended annotations.

My concern is the analysis tool specific comments going stale.  With cryptic one liners, they'd remain meaningless.  With the more detailed explanations, it wouldn't be so bad.

**Comment by mattcaswell** [comment]
The patches look relatively small and not too intrusive. I'd agree that fuller comments should be included. To be useful we'd definitely want to see this in our CI.

**Comment by Voker57** [comment]
> The patches look relatively small and not too intrusive. I'd agree that fuller comments should be included. To be useful we'd definitely want to see this in our CI.

One issue with including TF itself in your CI flow is it requires SYS_PTRACE capability on images running it, due to GDB usage. This capability is generally disabled on CI services, since it appears to allow leaking information about host processes.

**Comment by nhorman** [comment]
@Voker57 are you able to submit a PR for this?

**Comment by Voker57** [comment]
> @Voker57 are you able to submit a PR for this?

Yes, we could run CI much more easily with this and later possibly submit PRs to adjust the annotations.

**Comment by nhorman** [comment]
ok, assigning to you as a community issue, thank you!

**Comment by t8m** [comment]
As the PR adding such annotations did not appear for more than 1 year I am closing this issue.
